chore: Add Unit Tests for Configure Report to Select Scores or Raw Scores (M2-8036)

sultanofcardio opened the pull request:
### 📝 Description

🔗 [Jira Ticket M2-8036](https://mindlogger.atlassian.net/browse/M2-8036)

This PR adds unit tests for the changes introduced in #1935 and #1962. As a precursor for implementing these, I did a couple quick type-related fixes.

Finally, I also took this opportunity to remove the husk pre-commit hook. I confirmed that it is not being used in any build pipelines

### 📸 Screenshots

N/A

### 🪤 Peer Testing

`npm run test`

### ✏️ Notes

N/A

sultanofcardio committed:
Type more mocks

sultanofcardio committed:
Type default configs

The `defaultNumberSelectionConfig` has a timer property that it shouldn't have, based on the API types and the data in the database

sultanofcardio committed:
Type even more mocks

sultanofcardio committed:
Remove husky

sultanofcardio committed:
Write tests

sultanofcardio committed:
Fix broken test

sultanofcardio committed:
Account for removed timer property

farmerpaul commented on package-lock.json:
I'm guessing you didn't mean to remove this file…

farmerpaul replied:
Oops, I automatically assumed the file was removed based on GitHub not rendering the diff. You didn't remove it, my bad! 😅

farmerpaul commented on src/modules/Dashboard/features/Applet/Schedule/ImportSchedulePopup/ImportSchedulePopup.utils.test.tsx:
```diff
@@ -307,6 +307,8 @@ describe('prepareImportPayload', () => {
   test('returns prepared import payload with valid data', () => {
     const result = prepareImportPayload(mockUploadedEvents, mockAppletData, mockRespondentId);
 
+    const endOfCurrentYear = endOfYear(new Date()).getFullYear();
```
I find the semantics here a bit confusing. `endOfCurrentYear` is simply just the _current year_, as far as I can deduce, not the date representing the end of the current year. Then below, this year is being used via string interpolation to manually generate the date representing the end of the current year. In the end, the call to `endOfYear` above isn't actually doing anything meaningful.

This would make more sense to me:

```tsx
const currentYear = new Date().getFullYear();

…

{
  endDate: `${currentYear}-12-31`,
  …
}
```

I'd either do that, or continue using `endOfYear` but make use of all the date information it returns to generate a formatted date.

sultanofcardio replied:
This is fair. I did this at the last minute without thinking about it too much. I think I'll opt for the latter approach 🙏

farmerpaul commented on src/shared/mock.ts:
😍

farmerpaul commented on src/modules/Builder/features/ActivitySettings/ScoresAndReports/ScoreContent/ScoreContent.test.tsx:
```diff
@@ -39,43 +51,90 @@ const tableItems = [
+  });
+
   test('should render score', () => {
+    // @ts-expect-error Ignored temporarily
```
Perhaps you're already addressing this, but just wanted to point out that these `@ts-expect-error` lines can be omitted if you just include:
```tsx
  isStaticActive: false,
```
in the `commonProps` definition above.

sultanofcardio replied:
Thanks. I planned to address these as part of a follow-up PR where I do more type fixes. At the time, I didn't want to dig into figuring out what's required to make it compile

farmerpaul commented on src/modules/Builder/features/ActivitySettings/ScoresAndReports/ScoreContent/ScoreContent.test.tsx:
```diff
@@ -88,25 +147,487 @@ describe('ScoreContent', () => {
+        expect(retryOkBtn.textContent).toBe('Ok');
+        await userEvent.click(retryOkBtn);
+      } catch {
+        // If it's not showing, that's fine
```
Can you explain why this is fine? Is this unit test non-deterministic (can have multiple valid behaviours)? Or is it always expected to throw an error? If it's the latter, then I would opt to use `queryBy*` over `getBy*` to avoid throwing errors, allowing expected behaviour to be articulated a little more explicitly.

sultanofcardio replied:
It's somewhat non-deterministic. The change score ID popup is shown whenever a user tries to change the name of a score while the score's ID (which is generated based on its calculation type + name) is being referenced as a variable somewhere inside the score (e.g. in the message shown, conditional logic, etc).

When you run this test without this try-catch, it works only for the sum case because the score ID is already made based on sum, so there would be no potential change for which to prompt the user for confirmation.

I wrote the test like this because it depends on a mock that may change in a way that causes this popup to show (as it has since this test was written), but its visibility isn't important to the outcome of the test.

I can add this context to the test, or perhaps I can just override the mock locally to avoid this. Let me know what you think

farmerpaul replied:
Thanks for the clarification! I'd be happy if you could just add this context in the form of a comment, sounds good to me.

sultanofcardio replied:
As it turns out my mocking of some of the hooks used by the component was responsible for the change in the test outcome, which wasn't immediately clear either. So I opted to override the mock locally instead to avoid any confusion

farmerpaul reviewed (approved):
Just a few minor comments and clarification questions, but looks good overall!

sultanofcardio committed:
Fix ignored type errors

sultanofcardio committed:
Fix typo

sultanofcardio committed:
Fix end of current year impl

sultanofcardio committed:
Override report message in test

This test relies on there not being scoreId references in its message body, so I override it to make sure to UI flow behaves as expected.

When the message body contains a variable reference to the scoreId, an unimportant interstitial popup appears, breaking the flow of the test

sultanofcardio committed:
Merge branch 'develop' into chore/M2-8036-subscale-scores-additional-tests

# Conflicts:
#	src/modules/Dashboard/features/Applet/Schedule/ImportSchedulePopup/ImportSchedulePopup.utils.test.tsx
